### PR TITLE
Hopefully fixed smart pointer class.

### PR DESCRIPTION
**Pull request by jaltmayerpizzorno**

### CODE REVIEW

**Review by sternj — APPROVED** [review]

File: src/include/pyptr.h [hunk 1]
```diff
@@ -15,23 +15,25 @@ class PyPtr {
  public:
   PyPtr(O* o) : _obj(o) {}
 
-  O* operator->() { return _obj; }
+  PyPtr(const PyPtr& ptr) : _obj(ptr._obj) {
+    Py_IncRef((PyObject*)_obj);
+  }
 
-  operator O*() { return _obj; }
+   O* operator->() { return _obj; }
```
Comment (emeryberger) [inline src/include/pyptr.h:22]: Please add an explanatory comment here (I understand why you did this but it should be documented).

File: src/include/pyptr.h [hunk 2]
```diff
@@ -15,23 +15,25 @@ class PyPtr {
...
-  PyPtr& operator=(O* o) {
-    Py_DecRef((PyObject*)_obj);
-    _obj = o;
-    return *this;
-  }
+ explicit operator O*() { return _obj; }
 
-  PyPtr& operator=(PyPtr& ptr) {
+  PyPtr& operator=(const PyPtr& ptr) {
+    // increment first in case self-assigning
```
Comment (emeryberger) [inline src/include/pyptr.h:27]: Because it could get GC'd if not, right? (Please elaborate either way)

Comment (emeryberger) [reply src/include/pyptr.h:27]: Also, would an explicit equality check make sense instead?

**Review by emeryberger — CHANGES_REQUESTED** [review]

File: src/include/pyptr.h (hunk 2 above)
Comment (jaltmayerpizzorno) [reply src/include/pyptr.h:26]: Yes -- if assigning to oneself, if you first decrement, you may lose the object. I'll add the equality check.

Comment (sternj) [reply src/include/pyptr.h:26]: I mean an explicit check for equality with `nullptr` could make sense, but I think that the information that the `static_cast` provides to us as developers is helpful and is a good canary for if other changes will cause type issues

File: src/include/pyptr.h (hunk 1 above)
Comment (sternj) [reply src/include/pyptr.h]: I removed that operator actually, I just replaced it with a static cast

Comment (jaltmayerpizzorno) [reply src/include/pyptr.h]: Sam got rid of that, so I think it's resolved.

**Review by emeryberger — APPROVED** [review]
LGTM.

**Pull request merged**